paraatm/simulation_method/vcas.py

```diff
@@ -0,0 +1,380 @@
+                ac.setRocd_fps(vy[ind])
+                self.simulationInterface.resume(1)
+                # check pause
+                while True:
+                    server_runtime_sim_status = self.simulationInterface.get_runtime_sim_status()
+                    # print(server_runtime_sim_status)
+                    if (server_runtime_sim_status == self.NATS_SIMULATION_STATUS_PAUSE):
+                        break
+
+                ac = self.aircraftInterface.select_aircraft(aclist[0])
+                h = ac.getAltitude_ft()
+                curr_t = self.starttime + self.simulationInterface.get_curr_sim_time()
+
+            if ind < ncmd - 2:  # the previous n-1 command
+                # print(ind)
+                if curr_t < cmd_time:
+                    # print(cmd_level)
+                    dt = cmd_time - curr_t
+                    stat = ac.getFlight_phase()
+                    ac.setCruise_alt_ft(cmd_level)
+                    ac.setFlight_phase(11)  # cruise
+                    ac.setCruise_tas_knots(tas[ind])
+
+                    self.simulationInterface.resume(dt)
+                    # check pause
+                    while True:
+                        server_runtime_sim_status = self.simulationInterface.get_runtime_sim_status()
+                        # print(server_runtime_sim_status)
+                        if (server_runtime_sim_status == self.NATS_SIMULATION_STATUS_PAUSE):
+                            break
+
+                    ac = self.aircraftInterface.select_aircraft(aclist[0])
+                    ac.setFlight_phase(stat)
+            else:  # the last command
+                stat = ac.getFlight_phase()
+                ac.setCruise_alt_ft(cmd_level)
+                ac.setFlight_phase(11)  # cruise
+                self.simulationInterface.resume(275)
+                # check pause
+                while True:
+                    server_runtime_sim_status = self.simulationInterface.get_runtime_sim_status()
+                    # print(server_runtime_sim_status)
+                    if (server_runtime_sim_status == self.NATS_SIMULATION_STATUS_PAUSE):
+                        break
+
+                ac = self.aircraftInterface.select_aircraft(aclist[0])
+                # curlon = ac.getLongitude_deg()
+                # curlat = ac.getLatitude_deg()
+                ac.setFlight_phase(stat)
+
+        self.simulationInterface.resume()
+        # check end
+        while True:
+            server_runtime_sim_status = self.simulationInterface.get_runtime_sim_status()
+            if (server_runtime_sim_status == self.NATS_SIMULATION_STATUS_ENDED):
+                break
+
+        millis = int(round(time.time() * 1000))
+        output_filename = os.path.splitext(os.path.basename(__file__))[0] + "_" + str(millis) + ".csv"
+        self.write_output(output_filename)
+        track = read_nats_output_file(output_filename)
+        self.cleanup()
+
+        return track
```

Comment:
This is more of a question:  the design of `NatsSimulationWrapper` is really for the `__call__` method to handle extracting and returning the output data.  It might be possible to remove the code you have here from `simulation`, and let that be done by using `__call__` (or the `()` operator) instead of `simulation`.  The `__call__` method will automatically make a call to `write_output` and read it back in behind the scenes, so it wouldn't be necessary to worry about constructing the filename yourself.  There may be some nuances to your method that I don't understand, but just a suggestion.

Reply:
Thanks for reviewing details in my code. The simulation function is a simulation with input commands. In the later functions, self.make_anti_model to be specific, self.simulation is continuously called. 
Actually, this code may not need the simulation wrapper class as input, since the NATS constants and interfaces are retrieved in the self.__init__ by calling NatsEnvironment.start_jvm function. This way, the multiple calls to self.simulation avoided repeated jvm start and shutdown and need only one time connection to the NATS standalone server. 
I feel the NATS wrapper a descent method for calling the simulation function and it would be consistent should every contribution utilizes this framework. That is one reason I still included it in my code.
Please let me know if I should remove the simulation wrapper or it is okay to keep it.

Reply:
I think what I would suggest is:

- Keep your `__init__` code as is.  You need to start the JVM prior to retrieving some of the variables that you are using.  `NatsEnvironment.start_jvm` is safe to call multiple times.
- Remove the code at the end of `simulation` that writes output and returns the `DataFrame`.  All of this, including creating a temporary file name for the NATS output file, and calling `cleanup`, is done automatically by `__call__`.
- In `make_anti_model`, replace `track = self.simulation(new_cmd)` with `track = self(new_cmd)` or `track = self.__call__(new_cmd)`.

This should do the same thing that you have now, but it eliminates the redundant code for writing output at the end of your `simulation` method.  By default `__call__` will remove the NATS output file after it has been read back in.  If you want to keep the output file, you can use `track = self(new_cmd, output_file=str(millis)+'.csv')`